src/Components/Components/src/Routing/Router.cs

```diff
@@ -114,8 +125,33 @@ private static string StringUntilAny(string str, char[] chars)
                 : str.Substring(0, firstIndex);
         }
 
+        private void RefreshRouteTable()
+        {
+            var assemblies = AdditionalAssemblies == null ? new[] { AppAssembly } : new[] { AppAssembly }.Concat(AdditionalAssemblies);
+            var assembliesSet = new HashSet<Assembly>(assemblies);
+            if (_assemblies.Count != assembliesSet.Count)
+            {
+                Routes = RouteTableFactory.Create(assemblies);
+                _assemblies.UnionWith(assembliesSet);
+            }
+            // If the new assemblies set is the same length as the previous one,
+            // we check if they are equal-by-value and refresh the route table if
+            // not. NOTE: This is an uncommon occurrence and a cold-code path so
+            // we shouldn't end up needing to do an equal-by-value check too frequently.
+            if (_assemblies.Count == assembliesSet.Count)
+            {
+                if (!_assemblies.SetEquals(assembliesSet))
+                {
+                    Routes = RouteTableFactory.Create(assemblies);
+                    _assemblies.UnionWith(assembliesSet);
```

Comment:
Is the purpose of `UnionWith` to stop people from removing assemblies from the route table after they have been added?

This is probably fine either way since it would be really weird for someone to want to remove assemblies, however it's the sort of thing we should mention in docs. It might be slightly easier for people to understand if we did allow replacing the set outright rather than unioning with the previous version, but I don't mind strongly.

Reply:
I would say this is a rather weird behavior. We should not try to outsmart users and just give them what they asked for. If we have to explain it in the docs it will be a bad behavior to have on by default, and given how niche this is, I'd rather not have it and just stick with whatever assemblies the user set.

We will have docs and samples on how to do lazy loading, and the majority of our users will just follow our guidance.

Reply:
This wasn't intentional. We should be calling `_assemblies.Clear()` before adding the new set.